### DISCUSSION TITLE
Getting `Caps Lock` status :)

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#221] [Show and tell]
Back in 2018 I had a sketch in Processing Python mode where I wanted to use the `Caps Lock` status as an interface input, to change "view modes".

One can check for `key == CODED and key_code == 20` to see if `Caps Lock` was pressed, but then you have to keep track of the status in a global flag variable, etc., and I didn't want any of that... (and you actually can track changes but not the the initial status...) so I went digging and found `java.awt.Toolkit.getDefaultToolkit().getLockingKeyState(20)`.

You can still do that with py5! You just have to import `java` from `jpype`. As a bonus, while I was rewriting this example, I found out that `key` seems to start as `NUL` / `chr(0)` at the start of a sketch, good to know, I suppose...

```python
import py5
from java.awt import Toolkit

def setup():
    py5.size(400, 400)
    py5.text_size(40)
    py5.text_align(py5.CENTER, py5.CENTER)
    py5.text_font(py5.create_font('Noto Sans', 40))
def draw():
    if py5.frame_count == 1:
        print(repr(py5.key), ord(py5.key), py5.key_code) 
    py5.background(100)
    is_caps_locked = Toolkit.getDefaultToolkit().getLockingKeyState(20)
    if is_caps_locked:
        py5.no_fill()
        py5.rect(100, 100, 100, 100)
    py5.fill(0, 0, 100)    
    py5.text(py5.key, 150, 150)
    
def key_pressed(e):
    if py5.key == py5.CODED and py5.key_code == 20:
        print('CAPS LOCK')
    else:
        print(py5.key, py5.key_code)

py5.run_sketch()
```
![image](https://user-images.githubusercontent.com/3694604/212749931-2848ee34-eef7-4ed2-9d40-1a8544c54a34.png)

PS: The Noto fonts, named after the noble goal of "No Tofu", no missing glyphs, just doesn't live to its aspirations:  ASCII 20 "Device Control 4" has no glyph! :laughing: .

UPDATED: I updated with the tip bellow from @hx2A!!!

**Comment by hx2A** [comment]
Neat! Thank you for sharing this.

If you want to make it even more elegant, you can try this:

```python
import py5

from java.awt import Toolkit
```

JPype does some Python sorcery to make importing Java packages look like importing Python libraries. Anything on the classpath is applicable. py5 sets this up for you when it is imported (after it starts the JVM). You can then import `Toolkit` using the above code if you do it after `import py5`.

Happy coding!

**Reply by villares** [reply]
Great, thank you! I must have tried it in the wrong order, because I had a vague memory I could import stuff from java thanks to jpype but couldn't do it. I have updated the post now!
